File: app/tools/brush_tool.py
import numpy as np
import cv2
from PySide6.QtCore import QPointF, Qt
from PySide6.QtGui import QMouseEvent, QPainter, QPen, QColor
from app.tools.base_tool import BaseTool

class MaskBrushTool(BaseTool):
    """
    Brush/Eraser tool for painting directly onto the alpha mask.
    Mode: "Restore" (value=255) or "Eraser" (value=0).
    """

    def __init__(self, canvas, mode: str = "Restore"):
        super().__init__(canvas)
        self.mode = mode               # "Restore" or "Eraser"
        self.size = 30                 # Brush diameter in pixels
        self.hardness = 0.8            # 0.0 to 1.0
        self.opacity = 1.0             # 0.0 to 1.0
        self.is_drawing = False
        self.last_pos = None
        self.current_mask = None

    def set_mode(self, mode: str):
        self.mode = mode

    def mouse_press(self, img_pos: QPointF, event: QMouseEvent):
        if event.button() != Qt.MouseButton.LeftButton or not self.canvas.document:
            return

        doc = self.canvas.document
        lyr = doc.active_layer
        if not lyr or lyr.image is None:
            return

        self.is_drawing = True
        layer_pos = doc.map_canvas_pos_to_layer_pos(img_pos, lyr)
        self.last_pos = (int(round(layer_pos[0])), int(round(layer_pos[1])))

        h, w = lyr.height(), lyr.width()
        if lyr.mask is None:
            lyr.mask = np.full((h, w), 255, dtype=np.uint8)

        self.current_mask = lyr.mask.copy()
        self._paint_circle(self.last_pos, lyr)
        self.canvas.update()

    def mouse_move(self, img_pos: QPointF, event: QMouseEvent):
        doc = self.canvas.document
        lyr = doc.active_layer if doc else None
        if self.is_drawing and self.current_mask is not None and lyr:
            layer_pos = doc.map_canvas_pos_to_layer_pos(img_pos, lyr)
            curr_pos = (int(round(layer_pos[0])), int(round(layer_pos[1])))
            if self.last_pos is not None:
                self._paint_line(self.last_pos, curr_pos, lyr)
            self.last_pos = curr_pos
        self.canvas.update()

    def mouse_release(self, img_pos: QPointF, event: QMouseEvent):
        if event.button() == Qt.MouseButton.LeftButton and self.is_drawing:
            self.is_drawing = False
            doc = self.canvas.document
            lyr = doc.active_layer if doc else None
            if doc and lyr and self.current_mask is not None:
                description = "Restore Brush" if self.mode == "Restore" else "Erase Brush"
                doc.update_mask(self.current_mask, layer_id=lyr.id, description=description)
            self.current_mask = None
            self.last_pos = None
            self.canvas.update()

    def _paint_circle(self, center: tuple, lyr):
        if self.current_mask is None:
            return

        cx, cy = center
        scale = max(0.1, max(lyr.scale_x, lyr.scale_y))
        r = max(1, int(round((self.size / 2.0) / scale)))
        val = 255 if self.mode == "Restore" else 0

        cv2.circle(self.current_mask, (cx, cy), r, (val,), -1)

    def _paint_line(self, p1: tuple, p2: tuple, lyr):
        if self.current_mask is None:
            return

        scale = max(0.1, max(lyr.scale_x, lyr.scale_y))
        r = max(1, int(round((self.size / 2.0) / scale)))
        val = 255 if self.mode == "Restore" else 0

        cv2.line(self.current_mask, p1, p2, (val,), thickness=r * 2)


    def draw_overlay(self, painter: QPainter):
        """Draws brush radius preview ring around current mouse position."""
        if self.canvas.hover_img_pos is not None:
            pos = self.canvas.hover_img_pos
            painter.setRenderHint(QPainter.RenderHint.Antialiasing)

            pen_color = QColor(0, 255, 128) if self.mode == "Restore" else QColor(255, 64, 64)
            pen = QPen(pen_color, 1.5, Qt.PenStyle.SolidLine)
            painter.setPen(pen)
            painter.setBrush(Qt.BrushStyle.NoBrush)

            r = self.size / 2.0
            painter.drawEllipse(pos, r, r)

File: src/tools/BrushTool.cpp
#include "tools/BrushTool.h"
#include "ui/CanvasView.h"
#include "core/ImageDocument.h"
#include "core/Layer.h"
#include <opencv2/imgproc.hpp>
#include <QPen>
#include <QColor>
#include <cmath>

namespace ImageCut {
namespace Tools {

MaskBrushTool::MaskBrushTool(UI::CanvasView* canvas, const QString& mode)
    : BaseTool(canvas), m_mode(mode) {}

void MaskBrushTool::setMode(const QString& mode) {
    m_mode = mode;
}

void MaskBrushTool::mousePress(const QPointF& imgPos, QMouseEvent* event) {
    if (event->button() != Qt::LeftButton || !m_canvas || !m_canvas->getDocument()) return;

    auto doc = m_canvas->getDocument();
    auto lyr = doc->getActiveLayer();
    if (!lyr || lyr->image.empty()) return;

    m_isDrawing = true;
    auto [lx, ly] = doc->mapCanvasPosToLayerPos(imgPos, lyr);
    m_lastPos = cv::Point(static_cast<int>(std::round(lx)), static_cast<int>(std::round(ly)));

    int h = lyr->height();
    int w = lyr->width();
    if (lyr->mask.empty()) {
        lyr->mask = cv::Mat(h, w, CV_8UC1, cv::Scalar(255));
    }

    m_currentMask = lyr->mask.clone();
    paintCircle(m_lastPos, lyr);
    m_canvas->updateViewport();
}

void MaskBrushTool::mouseMove(const QPointF& imgPos, QMouseEvent* event) {
    (void)event;
    if (!m_canvas) return;
    auto doc = m_canvas->getDocument();
    auto lyr = doc ? doc->getActiveLayer() : nullptr;

    if (m_isDrawing && !m_currentMask.empty() && lyr) {
        auto [lx, ly] = doc->mapCanvasPosToLayerPos(imgPos, lyr);
        cv::Point currPos(static_cast<int>(std::round(lx)), static_cast<int>(std::round(ly)));
        paintLine(m_lastPos, currPos, lyr);
        m_lastPos = currPos;
        lyr->mask = m_currentMask;
        lyr->invalidateCache();
        m_canvas->updateViewport(true);
    } else {
        if (m_canvas->viewport()) m_canvas->viewport()->update();
    }
}

void MaskBrushTool::mouseRelease(const QPointF& imgPos, QMouseEvent* event) {
    (void)imgPos;
    if (event->button() == Qt::LeftButton && m_isDrawing) {
        m_isDrawing = false;
        if (m_canvas && m_canvas->getDocument()) {
            auto doc = m_canvas->getDocument();
            auto lyr = doc->getActiveLayer();
            if (lyr && !m_currentMask.empty()) {
                QString desc = (m_mode == "Restore") ? "Restore Brush" : "Erase Brush";
                doc->updateMask(m_currentMask, lyr->id, desc);
            }
        }
        m_currentMask.release();
        m_canvas->updateViewport();
    }
}

void MaskBrushTool::paintCircle(const cv::Point& center, std::shared_ptr<Core::Layer> lyr) {
    if (m_currentMask.empty() || !lyr) return;

    double scale = std::max(0.1, std::max(lyr->scaleX, lyr->scaleY));
    int r = std::max(1, static_cast<int>(std::round((m_size / 2.0) / scale)));
    uint8_t val = (m_mode == "Restore") ? 255 : 0;

    cv::circle(m_currentMask, center, r, cv::Scalar(val), -1);
}

void MaskBrushTool::paintLine(const cv::Point& p1, const cv::Point& p2, std::shared_ptr<Core::Layer> lyr) {
    if (m_currentMask.empty() || !lyr) return;

    double scale = std::max(0.1, std::max(lyr->scaleX, lyr->scaleY));
    int r = std::max(1, static_cast<int>(std::round((m_size / 2.0) / scale)));
    uint8_t val = (m_mode == "Restore") ? 255 : 0;

    cv::line(m_currentMask, p1, p2, cv::Scalar(val), r * 2);
}

void MaskBrushTool::drawOverlay(QPainter* painter) {
    if (!m_canvas || !painter) return;
    QPointF hoverPos = m_canvas->getHoverPos();
    if (hoverPos.x() >= 0 && hoverPos.y() >= 0) {
        painter->setRenderHint(QPainter::Antialiasing);
        QColor penColor = (m_mode == "Restore") ? QColor(0, 255, 128) : QColor(255, 64, 64);
        QPen pen(penColor, 1.5, Qt::SolidLine);
        painter->setPen(pen);
        painter->setBrush(Qt::NoBrush);

        double r = m_size / 2.0;
        painter->drawEllipse(hoverPos, r, r);
    }
}

} // namespace Tools
} // namespace ImageCut
